src/client/particles.cpp

```diff
@@ -38,102 +38,108 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 */
 
 Particle::Particle(
-	IGameDef *gamedef,
-	LocalPlayer *player,
-	ClientEnvironment *env,
-	const ParticleParameters &p,
-	const ClientTexRef& texture,
-	v2f texpos,
-	v2f texsize,
-	video::SColor color
-):
-	scene::ISceneNode(((Client *)gamedef)->getSceneManager()->getRootSceneNode(),
-		((Client *)gamedef)->getSceneManager()),
-	m_texture(texture)
+		IGameDef *gamedef,
+		LocalPlayer *player,
+		ClientEnvironment *env,
+		const ParticleParameters &p,
+		const ClientParticleTexRef &texture,
+		v2f texpos,
+		v2f texsize,
+		video::SColor color,
+		ParticleSpawner *parent,
+		std::unique_ptr<ClientParticleTexture> owned_texture
+	) :
+		scene::ISceneNode(((Client *)gamedef)->getSceneManager()->getRootSceneNode(),
+				((Client *)gamedef)->getSceneManager()),
+
+		m_expiration(p.expirationtime),
+
+		m_env(env),
+		m_gamedef(gamedef),
+		m_collisionbox(aabb3f(v3f(-p.size / 2.0f), v3f(p.size / 2.0f))),
+		m_texture(texture),
+		m_texpos(texpos),
+		m_texsize(texsize),
+		m_pos(p.pos),
+		m_velocity(p.vel),
+		m_acceleration(p.acc),
+		m_drag(p.drag),
+		m_jitter(p.jitter),
+		m_bounce(p.bounce),
+		m_player(player),
+		m_size(p.size),
+
+		m_base_color(color),
+		m_color(color),
+		m_collisiondetection(p.collisiondetection),
+		m_collision_removal(p.collision_removal),
+		m_object_collision(p.object_collision),
+		m_vertical(p.vertical),
+		m_animation(p.animation),
+		m_glow(p.glow),
+
+		m_parent(parent),
+		m_owned_texture(std::move(owned_texture))
 {
-	// Misc
-	m_gamedef = gamedef;
-	m_env = env;
-
-	// translate blend modes to GL blend functions
-	video::E_BLEND_FACTOR bfsrc, bfdst;
-	video::E_BLEND_OPERATION blendop;
-	const auto blendmode = texture.tex != nullptr
-			? texture.tex -> blendmode
-			: ParticleParamTypes::BlendMode::alpha;
-
-	switch (blendmode) {
-		case ParticleParamTypes::BlendMode::add:
-			bfsrc = video::EBF_SRC_ALPHA;
-			bfdst = video::EBF_DST_ALPHA;
-			blendop = video::EBO_ADD;
-		break;
-
-		case ParticleParamTypes::BlendMode::sub:
-			bfsrc = video::EBF_SRC_ALPHA;
-			bfdst = video::EBF_DST_ALPHA;
-			blendop = video::EBO_REVSUBTRACT;
-		break;
-
-		case ParticleParamTypes::BlendMode::screen:
-			bfsrc = video::EBF_ONE;
-			bfdst = video::EBF_ONE_MINUS_SRC_COLOR;
-			blendop = video::EBO_ADD;
-		break;
-
-		default: // includes ParticleParamTypes::BlendMode::alpha
-			bfsrc = video::EBF_SRC_ALPHA;
-			bfdst = video::EBF_ONE_MINUS_SRC_ALPHA;
-			blendop = video::EBO_ADD;
-		break;
-	}
+	// Set material
+	m_material = [&] {
```

Comment:
Why does this need a separate function? There are no `return` keywords needed in this part. You could as well put everything into a single code block which would make it less obscure.